File: implementations/fsharp/src/Skight.AgentPlatform.FSharp/McpSchemaTranslator.fs
namespace Skight.AgentPlatform.FSharp

open System
open System.Text.Json

module McpSchemaTranslator =

    /// Parses an MCP tools/list JSON-RPC response into a list of ToolSchema records
    let parseToolsListResponse (jsonResponse: string) : Result<ToolSchema list, string> =
        try
            use doc = JsonDocument.Parse(jsonResponse)
            let root = doc.RootElement
            let mutable errorElement = JsonElement()
            let mutable resultElement = JsonElement()

            if root.TryGetProperty("error", &errorElement) && errorElement.ValueKind <> JsonValueKind.Null then
                let mutable msgElement = JsonElement()
                let errMsg = if errorElement.TryGetProperty("message", &msgElement) then msgElement.GetString() else "Unknown JSON-RPC error"
                Error (sprintf "MCP Server returned error: %s" errMsg)
            elif not (root.TryGetProperty("result", &resultElement)) then
                Error "Invalid MCP response: Missing 'result' property"
            else
                let mutable toolsElement = JsonElement()
                if not (resultElement.TryGetProperty("tools", &toolsElement)) then
                    Ok []
                else
                    let schemas = [
                        for t in toolsElement.EnumerateArray() do
                            let mutable nameElem = JsonElement()
                            let mutable descElem = JsonElement()
                            let mutable schemaElem = JsonElement()

                            let nameStr = if t.TryGetProperty("name", &nameElem) then nameElem.GetString() else ""
                            let descStr = if t.TryGetProperty("description", &descElem) then descElem.GetString() else ""
                            let inputSchemaJson =
                                if t.TryGetProperty("inputSchema", &schemaElem) then
                                    schemaElem.GetRawText()
                                else
                                    """{"type":"object","properties":{}}"""

                            match ToolName.create nameStr with
                            | Ok name ->
                                yield {
                                    Name = name
                                    Description = descStr
                                    ParametersJson = inputSchemaJson
                                }
                            | Error _ -> ()
                    ]
                    Ok schemas
        with ex ->
            Error (sprintf "Failed to parse MCP tools/list response: %s" ex.Message)

    /// Parses an MCP tools/call JSON-RPC response into an execution result string
    let parseToolCallResponse (jsonResponse: string) : string =
        try
            use doc = JsonDocument.Parse(jsonResponse)
            let root = doc.RootElement
            let mutable errorElement = JsonElement()
            let mutable resultElement = JsonElement()

            if root.TryGetProperty("error", &errorElement) && errorElement.ValueKind <> JsonValueKind.Null then
                let mutable msgElement = JsonElement()
                let errMsg = if errorElement.TryGetProperty("message", &msgElement) then msgElement.GetString() else "Unknown JSON-RPC error"
                sprintf "Error from MCP server: %s" errMsg
            elif not (root.TryGetProperty("result", &resultElement)) then
                sprintf "Error from MCP server: Invalid response structure: %s" jsonResponse
            else
                let mutable isErrElem = JsonElement()
                let mutable contentElem = JsonElement()

                let isError = resultElement.TryGetProperty("isError", &isErrElem) && isErrElem.GetBoolean()

                if resultElement.TryGetProperty("content", &contentElem) then
                    let textList = [
                        for c in contentElem.EnumerateArray() do
                            let mutable textElem = JsonElement()
                            if c.TryGetProperty("text", &textElem) then
                                yield textElem.GetString()
                    ]
                    let combined = String.concat "\n" textList
                    if isError then sprintf "Error executing MCP tool:\n%s" combined
                    else combined
                else
                    resultElement.GetRawText()
        with ex ->
            sprintf "Error parsing MCP tool response: %s" ex.Message

File: implementations/csharp/src/Skight.AgentPlatform/McpSchemaTranslator.cs
using System;
using System.Collections.Generic;
using System.Text.Json;

namespace Skight.AgentPlatform
{
    public class McpToolSchema
    {
        public string Name { get; set; } = string.Empty;
        public string Description { get; set; } = string.Empty;
        public string ParametersJson { get; set; } = "{\"type\":\"object\",\"properties\":{}}";
    }

    public static class McpSchemaTranslator
    {
        public static (bool Success, List<McpToolSchema> Schemas, string Error) ParseToolsListResponse(string jsonResponse)
        {
            try
            {
                using var doc = JsonDocument.Parse(jsonResponse);
                var root = doc.RootElement;

                if (root.TryGetProperty("error", out var errorElem) && errorElem.ValueKind != JsonValueKind.Null)
                {
                    var errMsg = errorElem.TryGetProperty("message", out var msgElem) ? msgElem.GetString() ?? "Unknown JSON-RPC error" : "Unknown JSON-RPC error";
                    return (false, new List<McpToolSchema>(), $"MCP Server returned error: {errMsg}");
                }

                if (!root.TryGetProperty("result", out var resultElem))
                {
                    return (false, new List<McpToolSchema>(), "Invalid MCP response: Missing 'result' property");
                }

                var list = new List<McpToolSchema>();
                if (resultElem.TryGetProperty("tools", out var toolsElem) && toolsElem.ValueKind == JsonValueKind.Array)
                {
                    foreach (var t in toolsElem.EnumerateArray())
                    {
                        var nameStr = t.TryGetProperty("name", out var n) ? n.GetString() ?? "" : "";
                        var descStr = t.TryGetProperty("description", out var d) ? d.GetString() ?? "" : "";
                        var schemaJson = t.TryGetProperty("inputSchema", out var s) ? s.GetRawText() : "{\"type\":\"object\",\"properties\":{}}";

                        if (!string.IsNullOrWhiteSpace(nameStr))
                        {
                            list.Add(new McpToolSchema
                            {
                                Name = nameStr,
                                Description = descStr,
                                ParametersJson = schemaJson
                            });
                        }
                    }
                }

                return (true, list, string.Empty);
            }
            catch (Exception ex)
            {
                return (false, new List<McpToolSchema>(), $"Failed to parse MCP tools/list response: {ex.Message}");
            }
        }

        public static string ParseToolCallResponse(string jsonResponse)
        {
            try
            {
                using var doc = JsonDocument.Parse(jsonResponse);
                var root = doc.RootElement;

                if (root.TryGetProperty("error", out var errorElem) && errorElem.ValueKind != JsonValueKind.Null)
                {
                    var errMsg = errorElem.TryGetProperty("message", out var msgElem) ? msgElem.GetString() ?? "Unknown JSON-RPC error" : "Unknown JSON-RPC error";
                    return $"Error from MCP server: {errMsg}";
                }

                if (!root.TryGetProperty("result", out var resultElem))
                {
                    return $"Error from MCP server: Invalid response structure: {jsonResponse}";
                }

                var isError = resultElem.TryGetProperty("isError", out var errFlag) && errFlag.GetBoolean();

                if (resultElem.TryGetProperty("content", out var contentElem) && contentElem.ValueKind == JsonValueKind.Array)
                {
                    var texts = new List<string>();
                    foreach (var c in contentElem.EnumerateArray())
                    {
                        if (c.TryGetProperty("text", out var textElem))
                        {
                            texts.Add(textElem.GetString() ?? "");
                        }
                    }
                    var combined = string.Join("\n", texts);
                    return isError ? $"Error executing MCP tool:\n{combined}" : combined;
                }

                return resultElem.GetRawText();
            }
            catch (Exception ex)
            {
                return $"Error parsing MCP tool response: {ex.Message}";
            }
        }
    }
}
